Library/Homebrew/dependency.rb

```diff
@@ -71,25 +71,30 @@ class << self
     # The default filter, which is applied when a block is not given, omits
     # optionals and recommendeds based on what the dependent has asked for.
     def expand(dependent, deps = dependent.deps, &block)
+      @expand_stack ||= []
+      @expand_stack.push dependent
+
       expanded_deps = []
 
       deps.each do |dep|
-        # FIXME: don't hide cyclic dependencies
         next if dependent.name == dep.name
 
         case action(dependent, dep, &block)
         when :prune
           next
         when :skip
+          next if @expand_stack.include? dependent
```

Comment:
> Sorry, I'm not totally sure I understand how these two issues are related?

Let's assume we have a very small set of formulae A, B, and C where A depends on B and B depends on C:

- I would expect `brew uses C` to return B (the only direct user of C).
- I would expect `brew uses --recursive C` to return both B (the direct user of C) and A (uses C indirectly via B).

The same of course applies to a more complicated set of formulae and more levels of nesting in the dependency chains, but in the end `brew uses --recursive <formula>` should always return at least what `brew uses <formula>`. And then some more (all indirect users, too). Or am I misunderstanding some of the dependency stuff? The original code failed this simple check (I tried that locally with a random formula) and this is what I tried to get across in my comment.